### PR TITLE
Add table key_vault_managed_hardware_security_module closes #233

### LINKED ISSUE #233: Add table key_vault_managed_hardware_security_module
**References**
Add any related links that will help us understand the resource, including vendor documentation, related GitHub issues, and Go SDK documentation.

### PR DESCRIPTION
**Pull request by ParthaI**
# Integration test logs
<details>
  <summary>Logs</summary>
  
```
No env file present for the current environment:  staging 
 Falling back to .env config
No env file present for the current environment:  staging
customEnv TURBOT_TEST_EXPECTED_TIMEOUT undefined

SETUP: tests/azure_key_vault_managed_hardware_security_module []

PRETEST: tests/azure_key_vault_managed_hardware_security_module

TEST: tests/azure_key_vault_managed_hardware_security_module
Running terraform
data.azurerm_client_config.current: Refreshing state...
data.null_data_source.resource: Refreshing state...
azurerm_resource_group.named_test_resource: Creating...
azurerm_resource_group.named_test_resource: Creation complete after 3s [id=/subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest45865]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Creating...
azurerm_storage_account.named_test_resource: Creating...
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [10s elapsed]
azurerm_storage_account.named_test_resource: Still creating... [10s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [20s elapsed]
azurerm_storage_account.named_test_resource: Still creating... [20s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [30s elapsed]
azurerm_storage_account.named_test_resource: Still creating... [30s elapsed]
azurerm_storage_account.named_test_resource: Creation complete after 36s [id=/subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest45865/providers/Microsoft.Storage/storageAccounts/turbottest45865]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [40s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [50s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [1m0s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [1m10s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [1m20s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [1m30s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [1m40s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [1m50s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [2m0s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [2m10s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [2m20s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [2m30s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [2m40s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [2m50s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [3m0s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [3m10s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [3m20s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [3m30s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [3m40s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [3m50s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [4m0s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [4m10s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [4m20s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [4m30s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [4m40s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [4m50s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [5m0s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [5m10s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [5m20s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [5m30s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [5m40s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [5m50s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [6m0s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [6m10s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [6m20s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [6m30s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [6m40s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [6m50s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [7m0s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [7m10s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [7m20s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [7m30s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [7m40s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [7m50s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [8m0s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [8m10s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [8m20s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [8m30s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [8m40s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [8m50s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [9m0s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [9m10s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [9m20s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [9m30s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [9m40s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [9m50s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [10m0s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [10m11s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [10m21s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [10m31s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [10m41s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [10m51s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [11m1s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [11m11s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [11m21s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [11m31s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [11m41s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Still creating... [11m51s elapsed]
azurerm_key_vault_managed_hardware_security_module.named_test_resource: Creation complete after 11m51s [id=/subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest45865/providers/Microsoft.KeyVault/managedHSMs/turbottest45865]
azurerm_monitor_diagnostic_setting.named_test_resource: Creating...
azurerm_monitor_diagnostic_setting.named_test_resource: Still creating... [10s elapsed]
azurerm_monitor_diagnostic_setting.named_test_resource: Creation complete after 12s [id=/subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest45865/providers/Microsoft.KeyVault/managedHSMs/turbottest45865|turbottest45865]

Warning: Deprecated Resource

The null_data_source was historically used to construct intermediate values to
re-use elsewhere in configuration, the same can now be achieved using locals


Apply complete! Resources: 4 added, 0 changed, 0 destroyed.

Outputs:

object_id = 06fd46b0-a867-49a1-a4f1-f7768465caba
resource_aka = azure:///subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest45865/providers/Microsoft.KeyVault/managedHSMs/turbottest45865
resource_aka_lower = azure:///subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourcegroups/turbottest45865/providers/microsoft.keyvault/managedhsms/turbottest45865
resource_id = /subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest45865/providers/Microsoft.KeyVault/managedHSMs/turbottest45865
resource_name = turbottest45865
storage_account_id = /subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest45865/providers/Microsoft.Storage/storageAccounts/turbottest45865
subscription_id = d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8
tenant_id = cdffd708-7da0-4cea-abeb-0a4c334d7f64

Running SQL query: test-get-query.sql
[
  {
    "enable_soft_delete": true,
    "hsm_uri": "https://turbottest45865.managedhsm.azure.net/",
    "id": "/subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest45865/providers/Microsoft.KeyVault/managedHSMs/turbottest45865",
    "name": "turbottest45865",
    "region": "eastus",
    "resource_group": "turbottest45865",
    "soft_delete_retention_in_days": 8,
    "tenant_id": "cdffd708-7da0-4cea-abeb-0a4c334d7f64",
    "type": "Microsoft.KeyVault/managedHSMs"
  }
]
✔ PASSED

Running SQL query: test-hydrate-query.sql
[
  {
    "akas": [
      "azure:///subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest45865/providers/Microsoft.KeyVault/managedHSMs/turbottest45865",
      "azure:///subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourcegroups/turbottest45865/providers/microsoft.keyvault/managedhsms/turbottest45865"
    ],
    "name": "turbottest45865",
    "tags": {
      "Name": "turbottest45865"
    },
    "title": "turbottest45865"
  }
]
✔ PASSED

Running SQL query: test-list-query.sql
[
  {
    "id": "/subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest45865/providers/Microsoft.KeyVault/managedHSMs/turbottest45865",
    "name": "turbottest45865"
  }
]
✔ PASSED

Running SQL query: test-not-found-query.sql
null
✔ PASSED

Running SQL query: test-turbot-query.sql
[
  {
    "akas": [
      "azure:///subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest45865/providers/Microsoft.KeyVault/managedHSMs/turbottest45865",
      "azure:///subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourcegroups/turbottest45865/providers/microsoft.keyvault/managedhsms/turbottest45865"
    ],
    "name": "turbottest45865",
    "tags": {
      "Name": "turbottest45865"
    },
    "title": "turbottest45865"
  }
]
✔ PASSED

POSTTEST: tests/azure_key_vault_managed_hardware_security_module

TEARDOWN: tests/azure_key_vault_managed_hardware_security_module

SUMMARY:

1/1 passed.
```
</details>

# Example query results
<details>
  <summary>Results</summary>

### Basic Info
  
```sql
select
  name,
  id,
  hsm_uri,
  type,
  enable_soft_delete,
  region
from
  azure_key_vault_managed_hardware_security_module;
```

```
+-----------------+---------------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------
| name            | id                                                                                                                                          | hsm_uri                                   
+-----------------+---------------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------
| turbottest33853 | /subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest33853/providers/Microsoft.KeyVault/managedHSMs/turbottest33853 | https://turbottest33853.managedhsm.azure.n
+-----------------+---------------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------
```

### List soft delete enable HSM

```sql
select
  name,
  id,
  enable_soft_delete
from
  azure_key_vault_managed_hardware_security_module
where enable_soft_delete;
```

```
+-----------------+---------------------------------------------------------------------------------------------------------------------------------------------+--------------------+
| name            | id                                                                                                                                          | enable_soft_delete |
+-----------------+---------------------------------------------------------------------------------------------------------------------------------------------+--------------------+
| turbottest33853 | /subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest33853/providers/Microsoft.KeyVault/managedHSMs/turbottest33853 | true               |
+-----------------+---------------------------------------------------------------------------------------------------------------------------------------------+--------------------+
```

</details>

### CODE REVIEW

File: azure/table_azure_key_vault_managed_hardware_security_module.go [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+	return &plugin.Table{
+		Name:        "azure_key_vault_managed_hardware_security_module",
+		Description: "Azure Key Vault Managed Hardware Security Module",
+		Get: &plugin.GetConfig{
+			KeyColumns:        plugin.AllColumns([]string{"name", "resource_group"}),
+			Hydrate:           getKeyVaultManagedHardwareSecurityModule,
+			ShouldIgnoreError: isNotFoundError([]string{"ResourceNotFound", "ResourceGroupNotFound", "404"}),
+		},
+		List: &plugin.ListConfig{
+			Hydrate: listKeyVaultManagedHardwareSecurityModule,
```
Comment (khushboo9024) [inline azure/table_azure_key_vault_managed_hardware_security_module.go:27]:
```suggestion
			Hydrate: listKeyVaultManagedHardwareSecurityModules,
```

File: azure/table_azure_key_vault_managed_hardware_security_module.go [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+			},
+			{
+				Name:        "provisioning_state",
+				Description: "Provisioning state. Possible values include: 'ProvisioningStateSucceeded', 'ProvisioningStateProvisioning', 'ProvisioningStateFailed', 'ProvisioningStateUpdating', 'ProvisioningStateDeleting', 'ProvisioningStateActivated', 'ProvisioningStateSecurityDomainRestore', 'ProvisioningStateRestoring'",
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromField("Properties.ProvisioningState"),
+			},
+			{
+				Name:        "sku_family",
+				Description: "Contains SKU family name",
```
Comment (khushboo9024) [inline azure/table_azure_key_vault_managed_hardware_security_module.go:90]:
```suggestion
				Description: "Contains SKU family name.",
```

File: azure/table_azure_key_vault_managed_hardware_security_module.go [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromField("Properties.Sku.Family"),
+			},
+			{
+				Name:        "sku_name",
+				Description: "SKU name to specify whether the key vault is a standard vault or a premium vault",
```
Comment (khushboo9024) [inline azure/table_azure_key_vault_managed_hardware_security_module.go:96]:
```suggestion
				Description: "SKU name to specify whether the key vault is a standard vault or a premium vault.",
```

File: azure/table_azure_key_vault_managed_hardware_security_module.go [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromField("Properties.Sku.Name").Transform(transform.ToString),
+			},
+			{
+				Name:        "tenant_id",
+				Description: "The Azure Active Directory tenant ID that should be used for authenticating requests to the key vault",
```
Comment (khushboo9024) [inline azure/table_azure_key_vault_managed_hardware_security_module.go:102]:
```suggestion
				Description: "The Azure Active Directory tenant ID that should be used for authenticating requests to the key vault.",
```

File: azure/table_azure_key_vault_managed_hardware_security_module.go (hunk 2 above)
Comment (khushboo9024) [reply azure/table_azure_key_vault_managed_hardware_security_module.go:90]: Please check other places if period is missing.

File: azure/table_azure_key_vault_managed_hardware_security_module.go [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+			},
+			{
+				Name:        "enable_soft_delete",
+				Description: "Property to specify whether the 'soft delete' functionality is enabled for this managed HSM pool. If it's not set to any value(true or false) when creating new managed HSM pool, it will be set to true by default. Once set to true, it cannot be reverted to false.",
+				Type:        proto.ColumnType_BOOL,
+				Transform:   transform.FromField("Properties.EnableSoftDelete"),
+			},
+			{
+				Name:        "soft_delete_retention_in_days",
+				Description: "softDelete data retention days. It accepts >=7 and <=90.",
```
Comment (khushboo9024) [inline azure/table_azure_key_vault_managed_hardware_security_module.go:60]:
```suggestion
				Description: "Indicates softDelete data retention days. It accepts >=7 and <=90.",
```

File: azure/table_azure_key_vault_managed_hardware_security_module.go [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+				Transform:   transform.FromField("Properties.StatusMessage"),
+			},
+			{
+				Name:        "create_mode",
+				Description: "The create mode to indicate whether the resource is being created or is being recovered from a deleted resource. Possible values include: 'CreateModeRecover', 'CreateModeDefault'",
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromField("Properties.CreateMode"),
+			},
+			{
+				Name:        "provisioning_state",
```
Comment (khushboo9024) [inline azure/table_azure_key_vault_managed_hardware_security_module.go:83]: It should be the fourth column.

File: azure/table_azure_key_vault_managed_hardware_security_module.go [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+
+	var name, resourceGroup string
+	if h.Item != nil {
+		data := h.Item.(keyvault.ManagedHsm)
+		name = *data.Name
+		resourceGroup = strings.Split(*data.ID, "/")[4]
+	} else {
+		name = d.KeyColumnQuals["name"].GetStringValue()
+		resourceGroup = d.KeyColumnQuals["resource_group"].GetStringValue()
+	}
```
Comment (khushboo9024) [inline azure/table_azure_key_vault_managed_hardware_security_module.go:202]: Add check for empty `name` and `resource_group`.

File: azure/table_azure_key_vault_managed_hardware_security_module.go [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+		d.StreamListItem(ctx, vault)
+	}
+
+	return nil, err
+}
+
+//// HYDRATE FUNCTIONS
+
+func getKeyVaultManagedHardwareSecurityModule(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
+	plugin.Logger(ctx).Trace("getKeyVaultManagedHardwareSecurityService")
```
Comment (khushboo9024) [inline azure/table_azure_key_vault_managed_hardware_security_module.go:185]:
```suggestion
	plugin.Logger(ctx).Trace("getKeyVaultManagedHardwareSecurityModule")
```

File: azure/table_azure_key_vault_managed_hardware_security_module.go [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromField("ID").Transform(idToSubscriptionID),
+			},
+		},
+	}
+}
+
+//// LIST FUNCTION
+
+func listKeyVaultManagedHardwareSecurityModule(ctx context.Context, d *plugin.QueryData, _ *plugin.HydrateData) (interface{}, error) {
```
Comment (khushboo9024) [inline azure/table_azure_key_vault_managed_hardware_security_module.go:158]:
```suggestion
func listKeyVaultManagedHardwareSecurityModules(ctx context.Context, d *plugin.QueryData, _ *plugin.HydrateData) (interface{}, error) {
```

File: azure/table_azure_key_vault_managed_hardware_security_module.go [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+				Transform:   transform.FromField("Properties.Sku.Name").Transform(transform.ToString),
+			},
+			{
+				Name:        "tenant_id",
+				Description: "The Azure Active Directory tenant ID that should be used for authenticating requests to the key vault",
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromField("Properties.TenantID").Transform(transform.ToString),
+			},
+			{
+				Name:        "diagnostic_settings",
```
Comment (khushboo9024) [inline azure/table_azure_key_vault_managed_hardware_security_module.go:107]: `InitialAdminObjectIds` property is not added , please check.

File: docs/tables/azure_key_vault_managed_hardware_security_module.md [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+  id,
+  hsm_uri,
+  type,
+  enable_soft_delete,
+  region
+from
+  azure_key_vault_managed_hardware_security_module;
+```
+
+### List soft delete disabled hsm
```
Comment (khushboo9024) [inline docs/tables/azure_key_vault_managed_hardware_security_module.md:21]:
```suggestion
### List soft delete disabled hsm managed key vaults
```

**Review by khushboo9024 — APPROVED** [review]

**Review by bigdatasourav — APPROVED** [review]
LGTM

**Review by rajlearner17 — APPROVED** [review]
LGTM

**Pull request merged**